client/metalink.c

```diff
@@ -0,0 +1,433 @@
+/*
+ * Copyright (C) 2021 VMware, Inc. All Rights Reserved.
+ *
+ * Licensed under the GNU Lesser General Public License v2.1 (the "License");
+ * you may not use this file except in compliance with the License. The terms
+ * of the License are located in the COPYING file of this distribution.
+ */
+
+/*
+ * Module   : metalink.c
+ *
+ * Abstract :
+ *
+ *            tdnfclientlib
+ *
+ *            client library
+ *
+ * Authors  : Nitesh Kumar (kunitesh@vmware.com)
+ */
+
+#include "includes.h"
+
+static void
+TDNFMetalinkHashFree(
+    TDNF_ML_HASH_INFO *ml_hash_info
+    )
+{
+    if (!ml_hash_info)
+    {
+        return;
+    }
+
+    TDNF_SAFE_FREE_MEMORY(ml_hash_info->type);
+    TDNF_SAFE_FREE_MEMORY(ml_hash_info->value);
+    TDNF_SAFE_FREE_MEMORY(ml_hash_info);
+}
+
+static void
+TDNFMetalinkUrlFree(
+    TDNF_ML_URL_INFO *ml_url_info
+    )
+{
+    if (!ml_url_info)
+    {
+        return;
+    }
+
+    TDNF_SAFE_FREE_MEMORY(ml_url_info->protocol);
+    TDNF_SAFE_FREE_MEMORY(ml_url_info->type);
+    TDNF_SAFE_FREE_MEMORY(ml_url_info->location);
+    TDNF_SAFE_FREE_MEMORY(ml_url_info->url);
+    TDNF_SAFE_FREE_MEMORY(ml_url_info);
+}
+
+void
+TDNFMetalinkFree(
+    TDNF_ML_CTX *ml_ctx
+    )
+{
+    if (!ml_ctx)
+        return;
+
+    TDNF_SAFE_FREE_MEMORY(ml_ctx->filename);
+    TDNFDeleteList(&(ml_ctx->hashes),
+                      (TDNF_ML_FREE_FUNC)TDNFMetalinkHashFree);
+    TDNFDeleteList(&(ml_ctx->urls),
```

Comment:
redundant braces

Reply:
All comments has been taken care in the updated PR.